src/cluster.c

```diff
@@ -4942,6 +4964,131 @@ void addNodeReplyForClusterSlot(client *c, clusterNode *node, int start_slot, in
     setDeferredArrayLen(c, nested_replylen, nested_elements);
 }
 
+/* Add detailed information of a node to the output buffer of the given client. */
+void addNodeDetailsToShardReply(client *c, clusterNode *node) {
+    int reply_cnt = 0;
+    void *node_replylen = addReplyDeferredLen(c);
+    addReplyBulkCString(c, "id");
+    addReplyBulkCBuffer(c, node->name, CLUSTER_NAMELEN);
+    reply_cnt++;
+    int port = server.cluster_announce_port ? server.cluster_announce_port : server.port;
+    if (port) {
+        addReplyBulkCString(c, "port");
+        addReplyLongLong(c, node->port);
+        reply_cnt++;
+    }
+    int tls_port = server.cluster_announce_tls_port ? server.cluster_announce_tls_port : server.tls_port;
+    if (tls_port) {
+        addReplyBulkCString(c, "tls_port");
+        addReplyLongLong(c, tls_port);
+        reply_cnt++;
+    }
+    addReplyBulkCString(c, "ip");
+    addReplyBulkCString(c, node->ip);
+    reply_cnt++;
+    addReplyBulkCString(c, "endpoint");
+    addReplyBulkCString(c, getPreferredEndpoint(node));
+    reply_cnt++;
+    if (node->hostname) {
+        addReplyBulkCString(c, "hostname");
+        addReplyBulkCString(c, node->hostname);
+        reply_cnt++;
+    }
+    addReplyBulkCString(c, "role");
+    if (nodeIsSlave(node)) {
+        addReplyBulkCString(c, "replica");
+        addReplyBulkCString(c, "replication-offset");
+        addReplyLongLong(c, node->repl_offset);
+        reply_cnt+=2;
+    } else {
+        addReplyBulkCString(c, "master");
+        reply_cnt++;
+    }
+    addReplyBulkCString(c, "health");
+    const char *health_msg = NULL;
+    if (nodeFailed(node)) {
```

Comment:
I think our tenets here should be that the state is actionable for clients. So I think PFAIL (aka timedout) isn't that useful, since it's not committed yet.

Reply:
It could be actionable I think especially in the read-replica use case. Not saying I know of a client actually doing it, I can imagine though a latency-sensitive client skips PFAIL replicas for read traffic because otherwise there is a nontrivial amount of likelihood that a PFAIL replica is actually down permanently so knowing PFAIL would be useful if the client would like to reduce latency as much as possible, either when the cluster is in a steady state or going through some node failures. Since PFAIL is a well-known and well-defined state and exposing it is neither complicated nor costly, I would like to err on the side of exposing it and let the client get creative. My suggestion for including epoch follows the same philosophy.

Reply:
I don't agree we should err on the side of "as much of possible". I honestly think we should do the opposite, release as little as possible. My mentality is if we the maintainers of redis built our version of a perfect client, would we include this information, I would guess not. Transient failures does not imply we should avoid that node, I would rather wait for the consensus of the cluster to say the node is down.

Reply:
> I don't agree we should err on the side of "as much of possible". 

I too believe in the [KISS](https://en.wikipedia.org/wiki/KISS_principle) principle so to clarify a bit, I am not trying to push a big and bloated command. My concern is that exposing this state would require changing the public interface in the future, which carries backcompat risk (probably more so with epoch); given that it is such a trivial thing to add it now, that is why I'd like to bring this discussion up now.

>I would rather wait for the consensus of the cluster to say the node is down.

From a quick search, it seems like #10195 could benefit from this information.

Reply:
>  My concern is that exposing this state would require changing the public interface in the future, which carries backcompat risk (probably more so with epoch); given that it is such a trivial thing to add it now, that is why I'd like to bring this discussion up now.

This is one of the reasons we wanted to move to a dict, so that we can add more fields and clients can opt in to them. 

I pinged the tencent guy to weigh in. I would like to understand their use case a little bit more, but am more inclined to add it now. I'll add it to the top level comment as a consideration.

Reply:
SGTM. Thanks @madolson

Reply:
If we don't add it now, we can add it later under a different map key, or we can include another measure or indicator such as [CLUSTER COUNT-FAILURE-REPORTS](https://redis.io/commands/cluster-count-failure-reports) for each node.